blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
+			return err
+		}
+		if len(modifiedTreeNodes) > 0 {
+			updateTimeElapsed := time.Since(updateStartTime)
+			log.Debugf("Reachability reindex root updated to %s. "+
+				"Modified %d tree nodes and took %dms.",
+				dag.reachabilityReindexRoot.blockNode.hash,
+				len(modifiedTreeNodes), updateTimeElapsed.Milliseconds())
+			for modifiedTreeNode := range modifiedTreeNodes {
+				dag.reachabilityStore.setTreeNode(modifiedTreeNode)
+			}
+		}
+	}
+
 	return nil
 }
 
+func (dag *BlockDAG) updateReachabilityReindexRoot(newTreeNode *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	nextReindexRoot := dag.reachabilityReindexRoot
+	for {
+		candidateReindexRoot, modifiedNodes, found, err := dag.tryMovingReachabilityReindexRoot(nextReindexRoot, newTreeNode)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			break
+		}
+		for node := range modifiedNodes {
+			modifiedTreeNodes[node] = struct{}{}
+		}
+		nextReindexRoot = candidateReindexRoot
+	}
+
+	dag.reachabilityReindexRoot = nextReindexRoot
+	return modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tryMovingReachabilityReindexRoot(
+	reindexRoot *reachabilityTreeNode, newTreeNode *reachabilityTreeNode) (
+	newReindexRoot *reachabilityTreeNode, modifiedTreeNodes map[*reachabilityTreeNode]struct{}, found bool, err error) {
+
+	if !reindexRoot.isAncestorOf(newTreeNode) {
+		commonAncestor := reindexRoot.findCommonAncestor(newTreeNode)
+		return commonAncestor, nil, true, nil
+	}
+
+	chosenReindexRootChild, err := reindexRoot.findReachabilityTreeAncestorInChildren(newTreeNode)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	if newTreeNode.blockNode.blueScore-chosenReindexRootChild.blockNode.blueScore < reachabilityReindexWindow {
+		return nil, nil, false, nil
+	}
+	modifiedTreeNodes, err = dag.concentrateReachabilityTreeIntervalAroundReindexRootChild(reindexRoot, chosenReindexRootChild)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	return chosenReindexRootChild, modifiedTreeNodes, true, nil
+}
+
+// findReachabilityTreeAncestorInChildren finds the reachability tree child
+// of rtn that is the ancestor of node.
+func (rtn *reachabilityTreeNode) findReachabilityTreeAncestorInChildren(node *reachabilityTreeNode) (*reachabilityTreeNode, error) {
+
+	rootChildrenFutureCoveringSet := futureCoveringBlockSetFromReachabilityTreeNodes(rtn.children)
+	i := rootChildrenFutureCoveringSet.findIndex(&futureCoveringBlock{blockNode: node.blockNode, treeNode: node})
+	if i == 0 {
+		return nil, errors.Errorf("rtn is not an ancestor of node")
+	}
+
+	return rootChildrenFutureCoveringSet[i-1].treeNode, nil
+}
+
+func (dag *BlockDAG) concentrateReachabilityTreeIntervalAroundReindexRootChild(
+	reindexRoot *reachabilityTreeNode, chosenReindexRootChild *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
```

Comment:
It's a bit hard for a casual reader to understand what's the diff between `modifiedTreeNodes` and the two `modifiedNodes`s.
I'd suggest renaming to `allModifiedTreeNodes`, `beforeModifiedTreeNodes` and `afterModifiedTreeNodes` or something